lib/user/lookup.go

```diff
@@ -0,0 +1,237 @@
+	providers: []LookupProvider{
+		goLookupProvider{},
+		newGetentLookupProvider(),
+	},
+}
+
+type LookupProvider interface {
+	Lookup(username string) (*User, error)
+	LookupId(uid string) (*User, error)
+	LookupGroup(groupname string) (*Group, error)
+	LookupGroupId(gid string) (*Group, error)
+}
+
+func Lookup(username string) (*User, error) {
+	return DefaultLookupProvider.Lookup(username)
+}
+
+func LookupId(uid string) (*User, error) {
+	return DefaultLookupProvider.LookupId(uid)
+}
+
+func LookupGroup(groupname string) (*Group, error) {
+	return DefaultLookupProvider.LookupGroup(groupname)
+}
+
+func LookupGroupId(gid string) (*Group, error) {
+	return DefaultLookupProvider.LookupGroupId(gid)
+}
+
+type chainLookupProvider struct {
+	providers []LookupProvider
+}
+
+func (c *chainLookupProvider) Lookup(username string) (*User, error) {
+	var lastErr error
+	for _, provider := range c.providers {
+		u, err := provider.Lookup(username)
+		if err == nil {
+			return u, nil
+		}
+		lastErr = err
+	}
+	return nil, lastErr
+}
+
+func (c *chainLookupProvider) LookupId(uid string) (*User, error) {
+	var lastErr error
+	for _, provider := range c.providers {
+		u, err := provider.LookupId(uid)
+		if err == nil {
+			return u, nil
+		}
+		lastErr = err
+	}
+	return nil, lastErr
+}
+
+func (c *chainLookupProvider) LookupGroup(groupname string) (*Group, error) {
+	var lastErr error
+	for _, provider := range c.providers {
+		g, err := provider.LookupGroup(groupname)
+		if err == nil {
+			return g, nil
+		}
+		lastErr = err
+	}
+	return nil, lastErr
+}
+
+func (c *chainLookupProvider) LookupGroupId(gid string) (*Group, error) {
+	var lastErr error
+	for _, provider := range c.providers {
+		g, err := provider.LookupGroupId(gid)
+		if err == nil {
+			return g, nil
+		}
+		lastErr = err
+	}
+	return nil, lastErr
+}
+
+type goLookupProvider struct {
+}
+
+func (goLookupProvider) Lookup(username string) (*User, error) {
+	return gouser.Lookup(username)
+}
+
+func (goLookupProvider) LookupId(uid string) (*User, error) {
+	return gouser.LookupId(uid)
+}
+
+func (goLookupProvider) LookupGroup(groupname string) (*Group, error) {
+	return gouser.LookupGroup(groupname)
+}
+
+func (goLookupProvider) LookupGroupId(gid string) (*gouser.Group, error) {
+	return gouser.LookupGroupId(gid)
+}
+
+type getentLookupProvider struct {
+	getentPath string
+}
+
+func newGetentLookupProvider() getentLookupProvider {
+	getentPath := "getent"
+	if p := os.Getenv("CHEF_AUTOMATE_GETENT_PATH"); p != "" {
+		getentPath = p
+	} else {
+		// Find a system getent
+		for _, d := range []string{"/bin", "/usr/bin"} {
+			p := filepath.Join(d, "getent")
+			d, err := os.Stat(p)
+			if err != nil {
+				continue
+			}
+			if m := d.Mode(); !m.IsDir() && m&0111 != 0 {
+				getentPath = p
+				break
+			}
+		}
+	}
+
+	return getentLookupProvider{
+		getentPath: getentPath,
+	}
+}
+
+func (g getentLookupProvider) Lookup(username string) (*User, error) {
```

Comment:
not required at all for this PR, but from here down in this file there are a number of opportunities to pull out some well named constants and potentially extract some common code.  But I think it is fine as-is as well.

Reply:
that was the hope, but the error type got in the way